ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -590,140 +590,270 @@ private boolean copyDataByIndexSegment(LogSegment logSegmentToCopy, IndexSegment
+    return new FileSpan(start, end).inSpan(offset);
+  }
+
+  /**
+   * Adds entries related to {@code entry} that are in the same {@code indexSegment} including {@code entry}
+   * @param entries the list of {@link IndexEntry} to add to.
+   * @param indexSegment the {@link IndexSegment} to fetch values from.
+   * @param entry the {@link IndexEntry} for a DELETE that is under processing
+   * @throws StoreException if there are problems using the index
+   */
+  private void addAllEntriesForKeyInSegment(List<IndexEntry> entries, IndexSegment indexSegment, IndexEntry entry)
+      throws StoreException {
+    logger.trace("Fetching related entries of a blob with entry {} in index segment with start offset {} in {} "
+        + "because they need to be retained", entry, indexSegment.getStartOffset(), storeId);
+    NavigableSet<IndexValue> values = indexSegment.find(entry.getKey());
+    if (values.size() > 1) {
+      // we are using a multivalued index segment. Any related values will be in this set
+      values.forEach(valueFromSeg -> entries.add(new IndexEntry(entry.getKey(), valueFromSeg)));
+    } else {
+      // TODO: find a way to test this?
+      // in a non multi valued segment, there can only be PUTs and DELETEs in the same segment
+      entries.add(entry);
+      if (entry.getValue().isFlagSet(IndexValue.Flags.Delete_Index)) {
+        // this approach will fetch the PUT record.
+        IndexValue value = entry.getValue();
+        long putRecordOffset = value.getOriginalMessageOffset();
+        if (putRecordOffset != IndexValue.UNKNOWN_ORIGINAL_MESSAGE_OFFSET && putRecordOffset != value.getOffset()
+            .getOffset() && indexSegment.getStartOffset().getOffset() <= putRecordOffset) {
+          try (BlobReadOptions options = srcIndex.getBlobReadInfo(entry.getKey(),
+              EnumSet.allOf(StoreGetOptions.class))) {
+            Offset offset = new Offset(indexSegment.getStartOffset().getName(), options.getOffset());
+            MessageInfo info = options.getMessageInfo();
+            IndexValue putValue =
+                new IndexValue(info.getSize(), offset, info.getExpirationTimeInMs(), info.getOperationTimeMs(),
+                    info.getAccountId(), info.getContainerId());
+            entries.add(new IndexEntry(entry.getKey(), putValue));
+          } catch (StoreException e) {
+            logger.error("Fetching PUT index entry of {} in {} failed", entry.getKey(), indexSegment.getStartOffset());
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * @param key the {@link StoreKey} to check
+   * @param searchStartOffset the start offset of the search for delete entry
+   * @param searchEndOffset the end offset of the search for delete entry
+   * @return {@code true} if the key has been deleted
+   * @throws StoreException if there are any problems using the index
+   */
+  private boolean isDeleted(StoreKey key, Offset searchStartOffset, Offset searchEndOffset) throws StoreException {
```

Comment:
Add more comments why a put can be deleted, mention the fact that delete may not in current compaction range. 

Rename it to `hasDeleteEntryNotUnderCompaction` ?

Reply:
Made the changes discussed offline